### PR TITLE
Reaction Roles Revamp and move to discord-components

### LINKED ISSUE #39: More flexible role menu
Class categories should not be defined by a regex, but by the class list json. Destroy server function will need to be remade. Want to be able to handle more than just `XY 1234`

### LINKED ISSUE #41: Startup sequence is messy
Particularly server building. @CodingPenguin1 knows what to do. Needs to be brought into a single loop

### LINKED ISSUE #43: Move unique functions out of utils.py
Server management stuff should go to ServerManagement cog etc

### PR DESCRIPTION
**Pull request by CodingPenguin1**
Resolves #39 
Resolves #43 
Resolves #41 

EVERYTHING WITH ROLE MENUS WORKS

Added REQUIREMENTS.txt to list all requirements

-buildserver, -destroyserver, and -rolemenu are all cleaner. Built off CSV instead of JSON:
```
text,emoji,role/link,long_name,create_channels
CEG 3310,,CEG3310,some ceg class idk,"#ceg3310,0#Student Voice,2#TA Voice"
CEG 4324,,CEG4324,some other ceg class idk,"#ceg4324,0#Student Voice,2#TA Voice"
WSU Esports!,,https://discord.gg/smmV38d,,
CS 1000,,CS1000,demo cs class,"#cs1000,0#Student Voice,2#TA Voice"
EE 9999,,EE9999,demo ee class,"#ee9999,0#Student Voice,2#TA Voice"
```

Can specify channels per category, if starts with `#`, text channel. If starts with number before `#`, that's the max number of people allowed in the VC. 0 for infinite.

-buildserver and -rolemenu accept new csv as attachment. If passed, local copy replaced. Saved per-server.

Moved to discord-components instead of reactions for getting roles. Will need to be rewritten when discord-py 2.0 comes out, but that won't be for a while.

Role menus now function through a full bot restart or a cog restart. Cached in `role_menus.json`, which is updated every time a new role menu is made.

Lots more confirms for stuff

-rolemenu now automatically determines what channel to use first 2-3 characters of role column to determine which `*-class-selection` channel to use. If unsure, asks for user input.

Probably some other features, but I've forgotten them. Buttons are nice. Code is slightly bigger than old version I think, but it's _significantly_ more readable. Less caching in objects, and no more dicts of dicts of lists of dicts of dicts or whatever I was doing before

### CODE REVIEW

File: Cogs/ServerManagement.py [hunk 1]
```diff
@@ -14,131 +17,254 @@ class ServerManagement(commands.Cog):
...
+                chunk = buttons[i:i + 25]
+                menus.append([chunk[i:i + 5] for i in range(0, len(chunk), 5)])
+
+            # Clear channel
+            await channel.purge()
+
+            # Send each role menu
+            for menu in menus:
+                # Send and save message
+                message = await channel.send('‍', components=menu)  # 0 width joiner in here to send empty message
```
Comment (tux2603) [inline Cogs/ServerManagement.py:225]: Maybe use the escape code for the ZWJ?

**Review by tux2603 — CHANGES_REQUESTED** [review]

File: Cogs/ServerManagement.py [hunk 2]
```diff
@@ -14,131 +17,254 @@ class ServerManagement(commands.Cog):
...
+
+        # Get confirmation before building channels
+        if not await confirmation(self.bot, ctx, 'build'):
+            return
+
+        # Build all channels
+        for _, row in roles_csvs.iterrows():
+            # If role isn't a link, create role
+            if not validators.url(row['role/link']):
+                permissions = discord.Permissions(read_messages=True, send_messages=True, embed_links=True, attach_files=True, read_message_history=True, add_reactions=True, connect=True, speak=True, stream=True, use_voice_activation=True, change_nickname=True, mention_everyone=False)
```
[the remark this answers is not in the document]
Comment (CodingPenguin1) [reply Cogs/ServerManagement.py:64]: > 
> 
> This should get modified to fix issue #46

I'll move this to a separate PR

File: Cogs/ServerManagement.py (hunk 1 above)
Comment (CodingPenguin1) [reply Cogs/ServerManagement.py]: Resolved by 1d93ff4

**Comment by CodingPenguin1** [comment]
change ilocs in destroyserver() to df['header']

**Comment by CodingPenguin1** [comment]
Resolved #46

File: Cogs/ServerManagement.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodingPenguin1) [reply Cogs/ServerManagement.py:64]: Jk now resolved #46

File: Cogs/ServerManagement.py [hunk 3]
```diff
@@ -206,7 +207,8 @@ async def rolemenu(self, ctx):
 
                 # If role, make button style gray. If URL, make style URL
                 if not validators.url(role_link):
-                    buttons.append(Button(style=ButtonStyle.gray, label=text, emoji=emoji))
+                    # buttons.append(Button(style=ButtonStyle.gray, label=text, emoji=emoji))
+                    buttons.append(Button(style=ButtonStyle.gray, label=text, emoji=await get_emoji_named(ctx.guild, emoji)))
```
Comment (tux2603) [inline Cogs/ServerManagement.py:211]: Maybe add some sort of check to see if the server has an emoji with the name, and if not fall back to the default emojis?

**Review by tux2603 — CHANGES_REQUESTED** [review]

**Review by tux2603 — APPROVED** [review]

**Pull request merged**